keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,435 @@
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+The scale down constraints are a method to prevent too rapid loss of capacity. Hence, it makes sense to pick the maximum of two constraints.
+
+#### Stabilization Window
+
+Currently stabilization window ([PR][], [RFC][], [Algorithm Details][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and a new number of replicas is set to the maximum of all recommendations.
+
+It may be defined by command line option `--horizontal-pod-autoscaler-downscale-stabilization-window`.
+
+The same idea is used in the `DelaySeconds` parameter above:
+
+- While scaling up, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+- While scaling down, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
+
+The “stabilization window" as a result becomes an alias for the `constraints.scaleDown.delaySeconds`.
+
+[PR]: https://github.com/kubernetes/kubernetes/pull/68122
+[RFC]: https://docs.google.com/document/d/1IdG3sqgCEaRV3urPLA29IDudCufD89RYCohfBPNeWIM/edit#heading=h.3tdw2jxiu42f
+[Algorithm Details]: https://v1-14.docs.kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+
+#### API Changes
+
+The following API changes are needed:
+
+We should add `Scale{Up,Down}{Percent,Pods}` fields into the HPA spec
+
+The resulting solution will look like this:
+
+```golang
+type HPAScaleConstraintValue struct {
+    Rate         *HPAScaleConstraintRateValue
+    DelaySeconds *int32
+
+type HPAScaleConstraintRateValue struct {
+    Pods          *int32
+    Percent       *int32
+    PeriodSeconds *int32
+}
+
+type HPAScaleConstraints struct {
+    ScaleUp   *HPAScaleConstraintValue
+    ScaleDown *HPAScaleConstraintValue
+}
+
+type HorizontalPodAutoscalerSpec struct {
+    ScaleTargetRef CrossVersionObjectReference
+    MinReplicas    *int32
+    MaxReplicas    int32
+    Metrics        []MetricSpec
+    Constraints    *HPAScaleConstraints
+}
+```
+
+#### HPA Controller State Changes
+
+To store not only scale down recommendations, we need to replace
+
+```golang
+    recommendations map[string][]timestampedRecommendation
+```
+
+with
+
+```golang
+    scaleUpRecommendations   map[string][]timestampedRecommendation
+    scaleDownRecommendations map[string][]timestampedRecommendation
+```
+
+To store the information about last scale action, the HPA need an additional field (similar to the list of “recommendations”)
+
+```golang
+scaleEvents map[string][]timestampedScaleEvent
+```
+
+Where
+
+```golang
+type timestampedScaleEvent struct {
+    replicaChange int32
+    timestamp     time.Time
+}
+```
+
+It will store last scale events, and it will be used to make decisions about next scale events.
+
+Say, if 30 seconds ago the number of replicas was increased by one, and we forbid to scale up for more than 1 pod per minute,
```

Comment:
This looks right.  But I think it belongs up with the algorithm description.  This section is describing the specific state required to implement the algorithm.

Reply:
Moved to the "algorithm" section: https://github.com/kubernetes/enhancements/pull/883/commits/ec9c48e2a701c88b1afd88a204bc70a17aa9eb1f